public/docs/files/models/Configurely/Config.php.txt
<?php

namespace Configurely;

/**
 * This class provides a model for a Configurely application configuration.
 * 
 * @property-read int $id the unique identifier of the configuration
 * @property string $name the configuration name (non-unique)
 * @property string $description the configuration description
 * @property-read \DateTime $created_at the timestamp for when the configuration was created
 * @property-read \DateTime $updated_at the timestamp for when the configuration was last updated
 * @property-read string $app_link a uri to view the application to which the configuration belongs
 * @property-read string $settings_link a uri to list the settings for the configuration
*/
class Config extends Base {

    /**
     * provides a list of properties that should be included in the object's array and json output
    */
    protected $visible = array(
        'id',
        'name',
        'description',
        'created_at',
        'updated_at',
        'app_link',
        'settings_link'
    );

    /**
     * provides a list of custom properties that should be appended to the object's array and json output
    */
    protected $appends = array('app_link', 'settings_link');

    /**
     * provides a list of properties to explicitly allow for mass assignment
    */
    protected $fillable = array('name', 'description');

    /**
     * App (Eloquent)
     * the app to which the configuration belongs
    */
    public function app() {
        return $this->belongsTo('Configurely\App');
    }

    /**
     * Settings (Eloquent)
     * the list of settings for the configuration
    */
    public function settings() {
        return $this->hasMany('Configurely\Setting');
    }
    
    /**
     * App Link
     * gets the uri to view the application to which the configuration belongs
    */
    public function getAppLinkAttribute() {
        return \URL::action('AppController@show', $this->app->id);
    }

    /**
     * Settings Link
     * gets the uri to list the settings for the application
    */    
    public function getSettingsLinkAttribute() {
        return \URL::action('SettingController@index', [$this->app->id, $this->id]);
    }
    
    /**
     * gets the validation rules for the model
    */
    protected function rules() {
        return array_merge(
            parent::rules(),
            array(
                'name' => ['max:255', 'required'],
                'description' => 'max:4000'
            )
        );
    }
    
    /**
     * deletes the settings for the configuration and then deletes the configuration
    */
    public function delete() {
        $this->settings()->delete();
        return parent::delete();
    }
    
    /**
     * gets a configuration identified by the specified id and calls the delegate to get the response.
    */
    public static function getResponse($app_id, $config_id, $fn) {
        try {
			$config = Config::with('app', 'settings')->findOrFail($config_id);
		} catch(Exception $e) {
            return \ApiBaseController::getErrorResponse('ConfigController@index', [$app_id], \Lang::get('api.config_not_found'));
		}
        
        if($config->app->id != $app_id) {
            return \ApiBaseController::getErrorResponse('AppController@index', [], \Lang::get('api.bad_request'));
        }
        
        if(!\Session::get('apiuser')->owns($config->app)) {
            return \ApiBaseController::getErrorResponse('AppController@index', [$app_id], \Lang::get('api.app_not_authorized'));
        }
        
        return $fn($config);
    }
}


public/docs/files/controllers/ConfigController.php.txt
<?php

class ConfigController extends \ApiBaseController {

	/**
     * calls the Configurely\App::getResponse method to get a response containing the listing of configurations for the specified application or an error response indicating that the configurations can not be listed as requested
	 *
     * @param int $app_id the unique identifier of the application
     * 
	 * @return Response an html response if the request format is html, a json response otherwise
	 */
	public function index($app_id) {
        return Configurely\App::getResponse($app_id,
            function($app) {
                if($this->getRequestFormat() == 'html') {
                    return View::make('config.index')
                        ->with('app', $app);
                }
                return Response::json($app->configs);
            }
        );        
	}
    
	/**
     * calls the Configurely\App::getResponse method to get a response containing the form for creating a new configuration for the specified application or an error response indicating that the configuration can not be created
	 *
     * @param int $app_id the unique identifier of the application
     * 
	 * @return Response an html response
	 */
	public function create($app_id) {
        return Configurely\App::getResponse($app_id,
            function($app) {
                if($this->getRequestFormat() == 'html') {
                    return View::make('config.create')
                        ->with('app', $app);
                }
                // TODO: decide what to do if a json request is made to the create method
            }
        );
	}

	/**
	 * calls the Configurely\App::getResponse method to get a success response indicating that the configuration is stored successfully or an error response indicating that the configuration can not be stored
     * 
	 * @param int $app_id the unique identifier of the application
     *
	 * @return Response an html response if the request format is html, a json response otherwise
	 */
	public function store($app_id) {
        return Configurely\App::getResponse($app_id,
            function($app) {
                try {
                    $config = new Configurely\Config();
                    if(!$config->validate(Input::all())) {
                        return $this->getRedirectToCreate($app, $config->validator());
                    }
                    
                    $config->fill(Input::All());
                    $config->app()->associate($app);
                    $config->save();
                } catch(Exception $e) {
                    return $this->getErrorResponse('ConfigController@index', [$app->id], Lang::get('api.config_creation_failed'));
                }
                return $this->getSuccessResponse('ConfigController@index', [$app->id], Lang::get('api.config_created'));
            }
        );
	}
    
	/**
     * calls the Configurely\Config::getResponse method to get a response containing a view of the specified configuration or an error response indicating that the configuration can not be displayed
     *
	 * @param int $app_id the unique identifier of the application
     * @param int $config_id the unique identifier of the configuration
     * 
	 * @return Response an html response if the request format is html, a json response otherwise
	 */
	public function show($app_id, $config_id) {
        return Configurely\Config::getResponse($app_id, $config_id,
            function($config) {
                if($this->getRequestFormat() == 'html') {
                    return View::make('config.show')
                        ->with('config', $config);
                }
                return Response::json($config);
            }
        );
	}


	/**
     * calls the Configurely\Config::getResponse method to get a response containing the form for editing the specified configuration or an error response indicating that the configuration can not be edited
	 *
	 * @param int $app_id the unique identifier of the application
     * @param int $config_id the unique identifier of the configuration
     *
	 * @return Response an html response if the request format is html, a json response otherwise
	**/
	public function edit($app_id, $config_id) {
        return Configurely\Config::getResponse($app_id, $config_id,
            function($config) {
                if($this->getRequestFormat() == 'html') {
                    return View::make('config.edit')
                        ->with('config', $config);
                }
                // TODO: decide what to do if a json request is made to the edit method
            }
        );
	}

	/**
     * calls the Configurely\Config::getResponse method to get a success response indicating that the specified configuration is updated or an error response indicating that the configuration can not be updated
	 *
     * @param int $app_id the unique identifier of the application
     * @param int $config_id the unique identifier of the configuration
     *
	 * @return Response an html response if the request format is html, a json response otherwise
	 */
	public function update($app_id, $config_id) {
        return Configurely\Config::getResponse($app_id, $config_id,
            function($config) {
                try {
                    if(!$config->validate(Input::all())) {
                        return $this->getRedirectToEdit($config, $config->validator());
                    }
                    
                    $config->fill(Input::All());
                    $config->save();
                } catch(Exception $e) {
                    return $this->getErrorResponse('ConfigController@show', [$app_id, $config_id], Lang::get('api.config_update_failed'));
                }
                return $this->getSuccessResponse('ConfigController@show', [$app_id, $config_id], Lang::get('api.config_updated'));
            }
        );
	}

	/**
     * calls the Configurely\Config::getResponse method to get a success response indicating that the specified configuration is deleted or an error response indicating that the configuration can not be deleted
	 *
	 * @param int $app_id the unique identifier of the application
     * @param int $config_id the unique identifier of the configuration
     * 
	 * @return Response an html response if the request format is html, a json response otherwise
	 */
	public function destroy($app_id, $config_id) {
        return Configurely\Config::getResponse($app_id, $config_id,
            function($config) {
                try {
                    Configurely\Config::destroy($config->id);
                } catch(Exception $e) {
                    return $this->getErrorResponse('ConfigController@index', [$config->app->id], Lang::get('api.config_delete_failed'));
                }                
                return $this->getSuccessResponse('ConfigController@index', [$config->app->id], Lang::get('api.config_deleted'));
            }
        );
	}
    
    /**
     * gets a response containing a redirect to the create form with input and validation errors
     * 
     * @return Response an html response
     */
    protected function getRedirectToCreate($app, $validator) {
        return Redirect::action('ConfigController@create', [$app->id])
            ->withInput()
            ->withErrors($validator);
    }

    /**
     * gets a response containing a redirect to the edit form with input and validation errors
     *
     * @return Response an html response
     */
    protected function getRedirectToEdit($config, $validator) {
        return Redirect::action('ConfigController@edit', [$config->app->id, $config->id])
            ->withInput()
            ->withErrors($validator);
    }

}



app/models/Configurely/Config.php
<?php

namespace Configurely;

/**
 * This class provides a model for a Configurely application configuration.
 * 
 * @property-read int $id the unique identifier of the configuration
 * @property string $name the configuration name (non-unique)
 * @property string $description the configuration description
 * @property-read \DateTime $created_at the timestamp for when the configuration was created
 * @property-read \DateTime $updated_at the timestamp for when the configuration was last updated
 * @property-read string $app_link a uri to view the application to which the configuration belongs
 * @property-read string $settings_link a uri to list the settings for the configuration
*/
class Config extends Base {

    /**
     * provides a list of properties that should be included in the object's array and json output
    */
    protected $visible = array(
        'id',
        'name',
        'description',
        'created_at',
        'updated_at',
        'app_link',
        'settings_link'
    );

    /**
     * provides a list of custom properties that should be appended to the object's array and json output
    */
    protected $appends = array('app_link', 'settings_link');

    /**
     * provides a list of properties to explicitly allow for mass assignment
    */
    protected $fillable = array('name', 'description');

    /**
     * App (Eloquent)
     * the app to which the configuration belongs
    */
    public function app() {
        return $this->belongsTo('Configurely\App');
    }

    /**
     * Settings (Eloquent)
     * the list of settings for the configuration
    */
    public function settings() {
        return $this->hasMany('Configurely\Setting');
    }
    
    /**
     * App Link
     * gets the uri to view the application to which the configuration belongs
    */
    public function getAppLinkAttribute() {
        return \URL::action('AppController@show', $this->app->id);
    }

    /**
     * Settings Link
     * gets the uri to list the settings for the application
    */    
    public function getSettingsLinkAttribute() {
        return \URL::action('SettingController@index', [$this->app->id, $this->id]);
    }
    
    /**
     * gets the validation rules for the model
    */
    protected function rules() {
        return array_merge(
            parent::rules(),
            array(
                'name' => ['max:255', 'required'],
                'description' => 'max:4000'
            )
        );
    }
    
    /**
     * deletes the settings for the configuration and then deletes the configuration
    */
    public function delete() {
        $this->settings()->delete();
        return parent::delete();
    }
    
    /**
     * gets the breadcrumbs for the model
    */
    public function breadcrumbs() {
        $breadcrumbs = $this->app->breadcrumbs();
        if(count($breadcrumbs) > 0) {
            array_push($breadcrumbs, '/');
        }
        return array_merge(
            $breadcrumbs,
            array(
                ['Configurations' => \URL::action('ConfigController@index', $this->app->id)],
                '/',
                [e($this->name) => \URL::action('ConfigController@show', [$this->app->id, $this->id])]
            )
        );
    }
    
    /**
     * gets a configuration identified by the specified id and calls the delegate to get the response.
    */
    public static function getResponse($app_id, $config_id, $fn) {
        try {
			$config = Config::with('app', 'settings')->findOrFail($config_id);
		} catch(Exception $e) {
            return \ApiBaseController::getErrorResponse('ConfigController@index', [$app_id], \Lang::get('api.config_not_found'));
		}
        
        if($config->app->id != $app_id) {
            return \ApiBaseController::getErrorResponse('AppController@index', [], \Lang::get('api.bad_request'));
        }
        
        if(!\Auth::user()->owns($config->app)) {
            return \ApiBaseController::getErrorResponse('AppController@index', [$app_id], \Lang::get('api.app_not_authorized'));
        }
        
        return $fn($config);
    }
}